knimin/lib/format.py

```diff
@@ -0,0 +1,273 @@
+        The name of an additional contact person
+    contact_0_email : str
+        The email of an additional contact person
+    run_type : {"Target Gene", "Shotgun"}
+        Which data sheet structure to use
+    sample_information : iterable of dict
+        The sample information to load into the sample sheet. See note below.
+    contact_1_name : str, optional
+        The name of an additional contact.
+    contact_1_email : str, optional
+        The email of an additional contact.
+    contact_2_name : str, optional
+        The name of an additional contact.
+    contact_2_email : str, optional
+        The email of an additional contact.
+
+    Sample Sheet Note
+    -----------------
+    If the instrument type is a MiSeq, any lane information per-sample will be
+    disregarded. If instrument type is a HiSeq, each sample must include a
+    "lane" key.
+
+    If the run type is Target Gene, then sample details are disregarded
+    with the exception of determining the the lanes.
+
+    IF the run type is shotgun, then the following keys are required:
+        - sample-id
+        - i7-index-id
+        - i7-index
+        - i5-index-id
+        - i5-index
+
+    Raises
+    ------
+    ValueError
+        If a contact name is specified and an email is omitted, and vice versa
+    ValueError
+        If an unknown run type is specified
+    ValueError
+        If the number of cycles are <= 0
+    ValueError
+        If an unknown instrument type is specified.
+
+    Return
+    ------
+    str
+        The formatted sheet.
+    """
+    def validate_contact(name, email):
+        if name is not None and email is None:
+            raise ValueError("email is missing for %s" % name)
+        elif name is None and email is not None:
+            raise ValueError("name is missing for %s" % email)
+
+    c1name = contact_1_name if contact_1_name is not None else ''
+    c1email = contact_1_email if contact_1_email is not None else ''
+    c2name = contact_2_name if contact_2_name is not None else ''
+    c2email = contact_2_email if contact_2_email is not None else ''
+
+    validate_contact(pi_name, pi_email)
+    validate_contact(contact_0_name, contact_0_email)
+    validate_contact(contact_1_name, contact_1_email)
+    validate_contact(contact_2_name, contact_2_email)
+
+    if fwd_cycles <= 0 or not isinstance(fwd_cycles, int):
+        raise ValueError("fwd_cycles must be > 0")
+    if rev_cycles <= 0 or not isinstance(rev_cycles, int):
+        raise ValueError("rev_cycles must be > 0")
+
+    if run_type == 'Target Gene':
+        sample_header = DATA_TARGET_GENE_STRUCTURE
+        sample_detail = DATA_TARGET_GENE_SAMPLE_STRUCTURE
+    elif run_type == 'Shotgun':
+        sample_header = DATA_SHOTGUN_STRUCTURE
+        sample_detail = DATA_SHOTGUN_SAMPLE_STRUCTURE
+    else:
+        raise ValueError("%s is not a known run type" %
+                         run_type)
+
+    # if its a miseq, there isn't lane information
```

Comment:
Possibly. We only actually care about a few of the fields in here though, same with the instrument